sys: util: Improve IS_EMPTY macro

nordic-krch opened the pull request:
Replaced IS_EMPTY implementation with one found at
https://gustedt.wordpress.com/2010/06/08/detect-empty-macro-arguments/
which does not have limitation with regards to argument types. Former
implementation did not support arguments starting with brackets, dots,
quotations, etc.
    
Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
sys: util: Use IS_EMPTY macro implementation from P99

Replaced IS_EMPTY implementation with one from P99 which does not have
limitation with regards to argument types. Former implementation did
not support arguments starting with brackets, dots, quotations, etc.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
tests: unit: util: Extend IS_EMPTY test

Added various arguments to be tested against IS_EMPTY macro.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

pabigot commented:
Yes, this is the alternative I'd found too.  I'll answer [this](https://github.com/zephyrproject-rtos/zephyr/issues/26219#issuecomment-646487302) here:

The blog post has a clear copyright claim, and the incorporated work here appears derivative.  It is not trivial.  So a copyright acknowledgement may be appropriate.  No licensing is described.  So incorporating it this way is, in my non-lawyer opinion, problematic.

Reading the comments and searching further it appears this solution is available under QPI-1.0 in https://gforge.inria.fr/projects/p99/.   The specific function is in `<p99_args.h>`.  P99 also has comments that more clearly identify limitations on the functionality of this solution.

IMO it would be more appropriate to bring that project in as a module and start using it, which could allow removing much of the material in `sys/util.h` that has questionable provenance.

nordic-krch commented:
I agree with @pabigot that best would be to take P99 (QPL-1.0 license) as a module.  
@nashif can you comment? If yes, do we normally create fork in https://github.com/zephyrproject-rtos and use that?

de-nordic commented:
> 
> 
> I agree with @pabigot that best would be to take P99 (QPL-1.0 license) as a module.
> @nashif can you comment? If yes, do we normally create fork in https://github.com/zephyrproject-rtos and use that?

@pabigot , @nordic-krch  Create feature request or RFC for this as there is valid point in using the `p99` as it seems that we are deliberating on things that have been implemented there and are well documented. There are a lot of tests written too

carlescufi commented:
About p99, this uses the [QPL-1.0](https://opensource.org/licenses/QPL-1.0) license, which is OSI-approved but has not been approved for use in Zephyr, so this would have to go through the license approval process we have today. IANAL, but if I read this correctly, this requires redistributing the source code of any software built with this library:

>6. You may develop application programs, reusable components and other software items that link with the original or modified versions of the Software. These items, when distributed, are subject to the following requirements:

>a. You must ensure that all recipients of machine-executable forms of these items are also able to receive and use the complete machine-readable source code to the items without any charge beyond the costs of data transfer.

If my analysis is correct, this would be a no-go.

de-nordic commented:
> 
> If my analysis is correct, this would be a **complete no-go**.
 >
:(

nordic-krch commented:
So P99 has QPL so we won't take it but what about macros from the blog post? They are published there without the license, only copyright. In this particular case we are actually interested only in that macro.

carlescufi commented:
> So P99 has QPL so we won't take it but what about macros from the blog post? They are published there without the license, only copyright. In this particular case we are actually interested only in that macro.

The author of  the blog post is also an author of p99. I really don't have an answer for you, I don't know if we can take it.

carlescufi commented:
@nordic-krch I don't think we can take it freely:
https://softwareengineering.stackexchange.com/questions/148146/open-source-code-with-no-license-can-i-fork-it/148165#148165

carlescufi commented:
@gustedt would you be so kind as to weigh in in this Pull Request? Would you allow the code to be integrated into this header file under its current license? Thanks in advance.

gustedt commented:
Hi,
I'd prefer that you integrate P99 (partially or completely) than the code from the blog. There is no problem in broadening or specializing the license if need be, as long as this is one of the commonly use open source licenses.
Jens

pabigot commented:
@gustedt IMO that would be the ideal solution.  Could you add an option to p99 to license as Apache-2.0, or something compatible with Apache-2.0 or something compatible with it?

gustedt commented:
ok, I'll look into that. I have to move the project anyhow, because they will shut down the server on which it is located.

carlescufi commented:
> Hi,
> I'd prefer that you integrate P99 (partially or completely) than the code from the blog. There is no problem in broadening or specializing the license if need be, as long as this is one of the commonly use open source licenses.
> Jens

@gustedt 
Well, that would be wonderful, thank you for getting back to us. As you can see from the comment thread in this PR, QPL-1.0 is simply not acceptable in the context of Zephyr. Anything that is non-copyleft and accepted by the OSI should be good with us (eg. 3-clause BSD, Apache 2.0). Zephyr uses Apache 2.0, so this one is obviously the one we'd recommend. May I also suggest GitHub (or GitLab if you prefer) as a new home for the project.

gustedt commented:
I have now changed the license to Apache 2.0 and moved the git to the following location

https://gitlab.inria.fr/gustedt/p99

Please pull from there.
Jens

mbolivar-nordic commented:
> I have now changed the license to Apache 2.0

Awesome! It seems like you are not the sole author of this work:

https://gitlab.inria.fr/gustedt/p99/-/blob/master/p99/p99.h#L6

I assume you have obtained written consent from other authors for the license change?

gustedt commented:
> > I have now changed the license to Apache 2.0
> 
> Awesome! It seems like you are not the sole author of this work:
> 
> https://gitlab.inria.fr/gustedt/p99/-/blob/master/p99/p99.h#L6
> 
> I assume you have obtained written consent from other authors for the license change?

yes you may assume that

Jens

nordic-krch committed:
west.yml: Add P99 repository

Add P99 repository. P99 is a suite of macro and function definitions
that ease programming in modern C, aka C99.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch commented:
switched to use P99 but currently using remote repo directly and we need to make a fork in zephyrproject-rtos space. @carlescufi @nashif can you help with that?

pabigot reviewed (changes requested):
If we're adopting P99 (which I fully support) we should adopt it rather than hide it under a thin Zephyr wrapper.  `IS_EMPTY()` should be deprecated and all in-tree uses converted to `P99_IS_EMPTY()` in this PR.  The updated unit test should remain until the deprecated definition is removed in 2.6.

In a separate PR this conversion should be done for all other parts of `sys/util.h` that have P99 equivalents.  This would eliminate concern about the license and copyright of material that we could then remove.  It would also reduce the amount of namespace ambiguity and potential conflict with other external modules.

P99 is dual-licensed; is that an issue that requires TSC/Board/Linux Foundation input?  At a minimum I think the Zephyr repository should add a `README.Zephyr` that documents that use in this project is under the Apache-2.0 license.

gustedt commented:
> If we're adopting P99 (which I fully support) we should adopt it rather than hide it under a thin Zephyr wrapper. `IS_EMPTY()` should be deprecated and all in-tree uses converted to `P99_IS_EMPTY()` in this PR. The updated unit test should remain until the deprecated definition is removed in 2.6.

yes, I'd appreciate that

> In a separate PR this conversion should be done for all other parts of `sys/util.h` that have P99 equivalents. This would eliminate concern about the license and copyright of material that we could then remove. It would also reduce the amount of namespace ambiguity and potential conflict with other external modules.

and I'd appreciate this even more

Please, everybody working on this, don't hesitate to also propose patches to backport into P99 if there is need or occasion.

Jens

carlescufi commented:
@gustedt thank you very much for changing the license and moving to GitLab. I think we are now ready to formally proceed with the discussion of full p99 integration.

@pabigot @nordic-krch 
I suggest we do the following:

1. Update this PR with the Apache-2.0 licensed code from p99, just the snippet, in order to unblock progress required by use of this macro (@nordic-krch)
2. Create an issue about integrating p99 and include all the relevant stakeholders there (I will do that)

pabigot commented:
> 1. Update this PR with the Apache-2.0 licensed code from p99, just the snippet, in order to unblock progress required by use of this macro (@nordic-krch)
> 2. Create an issue about integrating p99 and include all the relevant stakeholders there (I will do that)

I disagree with with this plan, and don't see why the process proposed in https://github.com/zephyrproject-rtos/zephyr/pull/26299#pullrequestreview-439910676 can't be started now.  Add p99 as a module, and use it to redefine `IS_EMPTY()` to `P99_IS_EMPTY()` in the `include/sys/util*' headers.  The effort is not large, and we comply with the request of the project we're taking this from.

carlescufi commented:
> P99 is dual-licensed; is that an issue that requires TSC/Board/Linux Foundation input? At a minimum I think the Zephyr repository should add a `README.Zephyr` that documents that use in this project is under the Apache-2.0 license.



> > 1. Update this PR with the Apache-2.0 licensed code from p99, just the snippet, in order to unblock progress required by use of this macro (@nordic-krch)
> > 2. Create an issue about integrating p99 and include all the relevant stakeholders there (I will do that)
> 
> I disagree with with this plan, and don't see why the process proposed in [#26299 (review)](https://github.com/zephyrproject-rtos/zephyr/pull/26299#pullrequestreview-439910676) can't be started now. Add p99 as a module, and use it to redefine `IS_EMPTY()` to `P99_IS_EMPTY()` in the `include/sys/util*` headers. The effort is not large, and we comply with the request of the project we're taking this from.

The process you describe [requires an RFC or issue anyway](https://docs.zephyrproject.org/latest/guides/modules.html#process), so what I am doing is effectively starting the process. Since p99 is slightly different to other modules in the sense that it affects nearly every single platform and architecture, I named the issue RFC.

>P99 is dual-licensed; is that an issue that requires TSC/Board/Linux Foundation input? At a minimum I think the Zephyr repository should add a README.Zephyr that documents that use in this project is under the Apache-2.0 license.

I don't think so because, at least in the [p99](https://gitlab.inria.fr/gustedt/p99/-/tree/master/p99) folder, which might be the only one we import, the only license is Apache-2.0.

pabigot commented:
> in order to unblock progress required by use of this macro 

If #26219 is the sole motivation for this PR then nothing is blocked because I can do what I need with [the `FOR_EACH_NONEMPTY_TERM` macro]( https://github.com/zephyrproject-rtos/zephyr/pull/26062#issuecomment-647731800) which doesn't depend on changing `IS_EMPTY()`.

nordic-krch commented:
> FOR_EACH_NONEMPTY_TERM macro which doesn't depend on changing IS_EMPTY().

if your use case have arguments which can be concatenated ( does not start with `"`, etc.) then that will be enough.
